## MUSTDO!

- Tidy the code and increase readability THINK ABOUT STRUCTURE PRIOR. Create a test which compares the results to an output_file generated by using the ~/big_test.csv file as input just before beginning code clean up.



## TODO - TOP

- write tests for the newly implemented features in this branch (every single method/function should be tested)

- tidy up code

- Write some code that gets gene information at a given position from REST ensembl directly instead of relying on the bloated pyensembl module




## TODO - MINOR

- allow user to input there own transcript for scrapping exon information

- An option that allows you to use an annotated genome to get gene/transcript/exon info http://www.ensembl.org/info/data/ftp/index.html




## TESTING

The following AB1 files have given the expected output:
  B02_FUK20_VG_UKB_FUK20_F_004.ab1    A/G 15:48782270
  G03_HX9_AD_HX9_F_013.ab1            G 15:48787732
  D01_FUK31_LS_UKB_FUK31_F_007.ab1    C 17:48274593
  A01_HUK4_16_23DW_HUK4_              A/G 2:189863418
  B01_CXE_EM_CXE_F_003                W 17:48273298

The following AB1 files is giving 4 bases instead of the expected T/G. Good candidate for filtering bases by quality scores:
  H01_FUK15_RL_UKB_FUK15_F_015.ab1  A/G  15:48737567 

The following is a good match for multiple matching seq files:
  G03_HX9_AD_HX9_F_013.ab1   G 15:48787732

AB1 files where the _R.ab1 has the variant present but the _F.ab1 does not:
  C04_GXYUK2_RD_PP_GXYUK2_RD_R_006.ab1  G/C 2:189861900       post_seq
  HUK3-SC-R.ab1                         G/A    2:189861205

